File: 561.array-partition-i.cpp
/*
 * @lc app=leetcode id=561 lang=cpp
 *
 * [561] Array Partition I
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int arrayPairSum(vector<int>& nums) {
    std::sort(nums.begin(), nums.end());

    int size = nums.size();
    int ret = 0;
    for (int i = 0; i < size; i += 2) {
      ret += nums[i];
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  nums = {1, 4, 3, 2};
  ASSERT_EQ(s.arrayPairSum(nums), 4);
  nums = {6, 2, 6, 5, 1, 2};
  ASSERT_EQ(s.arrayPairSum(nums), 9);
  return 0;
}

File: 561.array-partition-i.rs
/*
 * @lc app=leetcode id=561 lang=rust
 *
 * [561] Array Partition I
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn array_pair_sum(mut nums: Vec<i32>) -> i32 {
        nums.sort_unstable();
        let mut ret = 0;
        for i in 0..(nums.len() / 2) {
            ret += nums[i * 2];
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::array_pair_sum(vec![1, 4, 3, 2]), 4);
}

#[test]
fn test2() {
    assert_eq!(Solution::array_pair_sum(vec![6, 2, 6, 5, 1, 2]), 9);
}
